### ISSUE TITLE
SNMP::INFO seems not threadsafe

### ISSUE THREAD

**Issue by ThomasJunk** [netdisco/snmp-info#431]
I am working on a threadded version of [SwitchMap](https://gitlab.com/ThomasJunk/switchmap).
When having different worker threads, each uses a SNMP::INFO object to populate an application object with data, the constructor seems to struggle with loading MIB-Information. Using it in a single thread everytihng is OK.

# Error
```Bash
Cannot find module (SNMPv2-MIB): At line 0 in (none)
Thread 3 terminated abnormally: The SNMPv2-MIB did not load. See README for SNMP::Info
```

or
```Bash
Cannot find module (IF-MIB)
```

The error is arbitrarily depending when the race codition happens.

# Expected
The MIBs should be loaded in a thread-safe manner without exception.

# Version
```Bash
cpan -D SNMP::Info
Reading $BLA
  Database was generated on Thu, 15 Jul 2021 13:29:03 GMT
SNMP::Info
-------------------------------------------------------------------------
	(no description)
	O/OL/OLIVER/SNMP-Info-3.73.tar.gz
	$BLA
	Installed: 3.73
	CPAN:      3.73  up to date
	Oliver Gorwits (OLIVER)
	oliver@cpan.org

```

# To reproduce

- set up 5 tun interfaces ('10.0.0.1', '10.0.0.2', '10.0.0.3','10.0.0.4', '10.0.0.5')
- use [SNMPSim](https://github.com/etingof/snmpsim) bound to said interfaces `sudo snmpsimd --data-dir=data --agent-udpv4-endpoint=10.0.0.5 --process-user=nobody --process-group=nogroup`

Alternatively: use 5 switches and their respective IPs

Run this test script:

```Perl
use strict;
use SNMP::Info 3.30;
use threads;
use threads::shared;
use Thread::Queue;


my $queue = Thread::Queue->new();
my @pool;

my @pool;
for (1 .. 5) {
    my $thr = threads->create(sub {
        while (my $name = $queue->dequeue()) {
                my $info = new SNMP::Info (
                                AutoSpecify => 1,
                                Debug       => 0, # note: setting this to 1 is interesting
                                DestHost    => $name,
                                Version     => 2
            );

        }
    });
    push @pool, $thr;
}

foreach ('10.0.0.1', '10.0.0.2', '10.0.0.3','10.0.0.4', '10.0.0.5') {
    $queue->enqueue($_);
}

for (1 .. 5) {
    $queue->enqueue(undef);
}

foreach my $thr (@pool) {
    $thr->join();
}
```

# Actual code
[The call to populate a Switch object](https://gitlab.com/ThomasJunk/switchmap/-/blob/threading/SwitchMap.pl#L510)
[The actual constructor gets called](https://gitlab.com/ThomasJunk/switchmap/-/blob/threading/Switch.pm#L264)

# Aside
As I am new to Perl and new to this Project perhaps I am overseeing something. 
Any help is appreciated. Thank you very much!

**Comment by JeroenvIS** [comment]
This is somewhat out of my comfort zone, but I'll give it a try.

SNMP::Info relies on SNMP.pm from the net-snmp package. SNMP.pm in turn uses the net-snmp C libraries.

According to [this net-snmp FAQ entry](https://net-snmp.sourceforge.io/wiki/index.php/FAQ:General_19) ([alt link](http://www.net-snmp.org/FAQ.html#Is_Net_SNMP_thread_safe_)) , net-snmp itself isn't thread safe, but can be used in a multithreaded application, referring to a README.thread document that can also be found [here](http://www.net-snmp.org/docs/README.thread.html).

Looks like in 2009 a patch was accepted for net-snmp 5.5 that improved thread safety for SNMP.pm, still with a few caveats; I found that [here](https://sourceforge.net/p/net-snmp/patches/907/).

Hope that this helps you or a more experienced developer in getting SNMP::Info working in a threaded application.

**Comment by ThomasJunk** [comment]
@JeroenvIS Thank you so far.
These documents confirm, what I found out so far: E.g. loading the MIBs isn't threadsafe. 
With your hints, I was able to make it work within the Switchmap enhancement. 
Having a threadsafe library would be the optimum. 
I was afraid, locking the actual call to the SNMP::INFO constructor would impact the performance gains from having multiple threads harder, but this is okay. 

Thank you for your patience and fast response. I'll close here for now.
